nvflare/app_opt/job_launcher/k8s_launcher.py

```diff
@@ -0,0 +1,277 @@
+
+    def launch_job(self, job_meta: dict, fl_ctx: FLContext) -> JobHandleSpec:
+
+        workspace_obj: Workspace = fl_ctx.get_prop(FLContextKey.WORKSPACE_OBJECT)
+        args = fl_ctx.get_prop(FLContextKey.ARGS)
+        client = fl_ctx.get_prop(FLContextKey.SITE_OBJ)
+        job_id = job_meta.get(JobConstants.JOB_ID)
+        server_config = fl_ctx.get_prop(FLContextKey.SERVER_CONFIG)
+        if not server_config:
+            raise RuntimeError(f"missing {FLContextKey.SERVER_CONFIG} in FL context")
+        service = server_config[0].get("service", {})
+        if not isinstance(service, dict):
+            raise RuntimeError(f"expect server config data to be dict but got {type(service)}")
+
+        self.logger.info(f"K8sJobLauncher start to launch job: {job_id} for client: {client.client_name}")
+        job_image = extract_job_image(job_meta, fl_ctx.get_identity_name())
+        self.logger.info(f"launch job use image: {job_image}")
+        job_config = {
+            "name": job_id,
+            "image": job_image,
+            "container_name": f"container-{job_id}",
+            "volume_mount_list": [{"name": self.workspace, "mountPath": self.mount_path}],
+            "volume_list": [{"name": self.workspace, "hostPath": {"path": self.root_hostpath, "type": "Directory"}}],
+            "module_args": {
+                "-m": args.workspace,
+                "-w": (workspace_obj.get_startup_kit_dir()),
+                "-t": client.token,
+                "-d": client.ssid,
+                "-n": job_id,
+                "-c": client.client_name,
+                "-p": "tcp://parent-pod:8004",
+                "-g": service.get("target"),
+                "-scheme": service.get("scheme", "grpc"),
+                "-s": "fed_client.json",
+            },
+            "set_list": args.set,
+        }
+
+        self.logger.info(f"launch job with k8s_launcher. Job_id:{job_id}")
+
+        job_handle = K8sJobHandle(job_id, self.core_v1, job_config, namespace=self.namespace, timeout=self.timeout)
+        try:
+            self.core_v1.create_namespaced_pod(body=job_handle.get_manifest(), namespace=self.namespace)
+            if job_handle.enter_states([JobState.RUNNING], timeout=self.timeout):
+                return job_handle
+            else:
+                job_handle.terminate()
+                return None
+        except ApiException as e:
+            job_handle.terminate()
+            return None
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.GET_JOB_LAUNCHER:
+            job_meta = fl_ctx.get_prop(FLContextKey.JOB_META)
+            job_image = extract_job_image(job_meta, fl_ctx.get_identity_name())
+            if job_image:
+                job_launcher: list = fl_ctx.get_prop(FLContextKey.JOB_LAUNCHER, [])
+                job_launcher.append(self)
+                fl_ctx.set_prop(FLContextKey.JOB_LAUNCHER, job_launcher, private=True, sticky=False)
```

Comment:
These 3 lines of code should be placed into a util func (in fed_utils or maybe in launcher_spec?), since it is used in all launcher impls.
add_launcher(launcher, fl_ctx)

Reply:
add add_launcher(launcher, fl_ctx) in the job_launcher_spec.py